File: python_test/classes.py
from time import time
from copy import deepcopy

import json

from dicttoxml import dicttoxml
import xmltodict

import yaml

import toml

class ParserABS:
    def __init__(self, struct_id, data, capacity, save=False, dtype='unknown'):
        self.data = data
        self.capacity = capacity
        self.stat = {
            "id": None,
            "struct_id":struct_id,
            "capacity": capacity,
            "serialize":None,
            "parse":None,
            "over":None,
        }
        self.archive = [
        ]
        self.save = save
        self.dtype = dtype

    def parse(self,data_str):
        data, timestamp = self._parse(data_str)
        self.stat["parse"] = timestamp/self.capacity
        self.stat["over"] = len(data_str)/self.capacity
        return data
    
    def serialize(self):
        data, timestamp = self._serialize(self.data)
        self.stat["serialize"] = timestamp/self.capacity
        # self.stat["over"].append(len(data_str)/self.capacity)
        return data

    def save_stat(self):
        self.archive.append(deepcopy(self.stat))

    def test(self, test_id):
        data_str = self.serialize()
        # print(data_str)
        # print("    ", self.capacity, len(data_str))
        data = self.parse(data_str)
        self.stat["id"] = test_id
        self.save_stat()
        return data, data_str

# ----------------------- JSON ---------------------------------
def json_parse(str_data):
    start = time()
    data = json.loads(str_data)
    stop = time()-start
    return data, stop

def json_serialize(data):
    start = time()
    str_data = json.dumps(data)
    stop = time()-start
    return str_data, stop


class ParserJSON(ParserABS):
    def  __init__(self, struct_id, data, capacity):
        super().__init__(struct_id, data, capacity)
        self._parse = json_parse
        self._serialize = json_serialize

# ----------------------- XML ---------------------------------

def xml_serialize(data):
    start = time()
    xml_str = dicttoxml(data, attr_type=False, root=False)
    stop = time()-start
    # print(xml_str)
    return "<root>"+xml_str.decode("utf-8")+"</root>", stop

def xml_parse(data_str):
    start = time()
    data = xmltodict.parse(data_str)
    stop = time()-start
    return data, stop

class ParserXML(ParserABS):
    def  __init__(self, struct_id, data, capacity):
        super().__init__(struct_id, data, capacity, save=False, dtype="xml")
        self._parse = xml_parse
        self._serialize = xml_serialize

# ----------------------- YAML --------------------------------

def yaml_serialize(data):
    start = time()
    yaml_str = yaml.dump(data)
    stop = time()-start
    return yaml_str, stop

def yaml_parse(data_str):
    start = time()
    data = yaml.load(data_str, Loader=yaml.BaseLoader)
    stop = time()-start
    return data, stop

class ParserYAML(ParserABS):
    def  __init__(self, struct_id, data, capacity):
        super().__init__(struct_id, data, capacity, save=False, dtype="yaml")
        self._parse = yaml_parse
        self._serialize = yaml_serialize

# ----------------------- TOML --------------------------------

def toml_serialize(data):
    start = time()
    yaml_str = toml.dumps(data)
    stop = time()-start
    return yaml_str, stop

def toml_parse(data_str):
    start = time()
    data = toml.loads(data_str)
    stop = time()-start
    return data, stop

class ParserTOML(ParserABS):
    def  __init__(self, struct_id, data, capacity):
        super().__init__(struct_id, data, capacity, save=False, dtype="toml")
        self._parse = toml_parse
        self._serialize = toml_serialize

File: js_test/classes.js
class ReaderABS{
    constructor(struct_id, capacity, data) {
        this._parse = null
        this._serialize = null
        this.data = data
        this.stat = {
            struct_id:struct_id,
            capacity:capacity,
            serialize:null,
            parse:null,
            over:null
        }
    }

    parse = (data_str)=>{
        var start = Date.now('nano')
        var data = this._parse(data_str)
        var stop = Date.now('nano') - start
        this.stat.parse = stop
    }

    serialize = ()=>{
        var start = Date.now('nano')
        var data_str = this._serialize(this.data)
        var stop = Date.now('nano') - start
        this.stat.serialize = stop
        this.stat.over = data_str.length/this.stat.capacity
    }

    save_stat = (path, test_id)=>{
        var content = test_id+","+this.stat.struct_id+","+this.stat.capacity+","+this.stat.serialize+","+this.stat.parse+","+this.stat.over
        fs.appendFile(path, content, err => {
            if (err) {
              console.error(err)
              return
            }
          })
    }

    test = (path, test_id) => {
        var data_str = this.serialize()
        this.parse(data_str)
        this.save_stat(path, test_id)
    }
}

// JSON
export class ReaderJSON extends ReaderABS{
    constructor(struct_id, capacity, data) {
        super(struct_id, capacity, data); 
        this._parse = JSON.parse
        this._serialize = JSON.stringify
    }
}

// XML
const xml_parser = require('xml2json');

export class ReaderXML extends ReaderABS{
    constructor(struct_id, capacity, data) {
        super(struct_id, capacity, data); 
        this._parse = (xml)=>{
            return xml_parser.toJson(xml, { object: true })
        }
        this._serialize = xml_parser.toXml
    }
}

// YAML
const yaml_parser = require('js-yaml');

export class ReaderYAML extends ReaderABS{
    constructor(struct_id, capacity, data) {
        super(struct_id, capacity, data); 
        this._parse = yaml_parser.load
        this._serialize = yaml_parser.dump
    }
}

// TOML
var toml_parser = require('@iarna/toml')
export class ReaderTOML extends ReaderABS{
    constructor(struct_id, capacity, data) {
        super(struct_id, capacity, data); 
        this._parse = toml_parser.parse
        this._serialize = toml_parser.stringify
    }
}